src/hotspot/cpu/x86/stubGenerator_x86_64_poly.cpp

```diff
@@ -0,0 +1,1036 @@
+  __ vpxorq(A1, A1, A1, Assembler::AVX_512bit);
+  __ vpxorq(A2, A2, A2, Assembler::AVX_512bit);
+  __ vpxorq(A3, A3, A3, Assembler::AVX_512bit);
+  __ vpxorq(A4, A4, A4, Assembler::AVX_512bit);
+  __ vpxorq(A5, A5, A5, Assembler::AVX_512bit);
+  __ vpxorq(B0, B0, B0, Assembler::AVX_512bit);
+  __ vpxorq(B1, B1, B1, Assembler::AVX_512bit);
+  __ vpxorq(B2, B2, B2, Assembler::AVX_512bit);
+  __ vpxorq(B3, B3, B3, Assembler::AVX_512bit);
+  __ vpxorq(B4, B4, B4, Assembler::AVX_512bit);
+  __ vpxorq(B5, B5, B5, Assembler::AVX_512bit);
+  __ vpxorq(R0, R0, R0, Assembler::AVX_512bit);
+  __ vpxorq(R1, R1, R1, Assembler::AVX_512bit);
+  __ vpxorq(R2, R2, R2, Assembler::AVX_512bit);
+  __ vpxorq(R1P, R1P, R1P, Assembler::AVX_512bit);
+  __ vpxorq(R2P, R2P, R2P, Assembler::AVX_512bit);
+  __ evmovdquq(Address(rsp, 64*3), A0, Assembler::AVX_512bit);
+  __ evmovdquq(Address(rsp, 64*4), A0, Assembler::AVX_512bit);
+  __ evmovdquq(Address(rsp, 64*5), A0, Assembler::AVX_512bit);
+  __ evmovdquq(Address(rsp, 64*0), A0, Assembler::AVX_512bit);
+  __ evmovdquq(Address(rsp, 64*1), A0, Assembler::AVX_512bit);
+  __ evmovdquq(Address(rsp, 64*2), A0, Assembler::AVX_512bit);
+  __ addq(rsp, 512/8*6); // (powers of R)
+}
+
+// This function consumes as many whole 16-byte blocks as available in input
+// After execution, input and length will point at remaining (unprocessed) data
+// and accumulator will point to the current accumulator value
+address StubGenerator::generate_poly1305_processBlocks() {
+  __ align(CodeEntryAlignment);
+  StubCodeMark mark(this, "StubRoutines", "poly1305_processBlocks");
+  address start = __ pc();
+  __ enter();
+
+  // Save all 'SOE' registers
+  __ push(rbx);
+  #ifdef _WIN64
+  __ push(rsi);
+  __ push(rdi);
+  #endif
+  __ push(r12);
+  __ push(r13);
+  __ push(r14);
+  __ push(r15);
+
+  // void processBlocks(byte[] input, int len, int[5] a, int[5] r)
+  const Register input        = rdi; //input+offset
+  const Register length       = rbx;
+  const Register accumulator  = rcx;
+  const Register R            = r8;
+
+  #ifdef _WIN64
+  // c_rarg0 - rcx
+  // c_rarg1 - rdx
+  // c_rarg2 - r8
+  // c_rarg3 - r9
+  __ mov(input, c_rarg0);
+  __ mov(length, c_rarg1);
+  __ mov(accumulator, c_rarg2);
+  __ mov(R, c_rarg3);
+  #else
+  // c_rarg0 - rdi
+  // c_rarg1 - rsi
+  // c_rarg2 - rdx
+  // c_rarg3 - rcx
+  // dont clobber R, args copied out-of-order
+  __ mov(length, c_rarg1);
+  __ mov(R, c_rarg3);
+  __ mov(accumulator, c_rarg2);
+  #endif
+
+  const Register a0 = rsi;  // [in/out] accumulator bits 63..0
+  const Register a1 = r9;   // [in/out] accumulator bits 127..64
+  const Register a2 = r10;  // [in/out] accumulator bits 195..128
+  const Register r0 = r11;  // R constant bits 63..0
+  const Register r1 = r12;  // R constant bits 127..64
+  const Register c1 = r8;   // 5*R (upper limb only)
+
+  Label L_process16Loop, L_process16LoopDone;
+
+  // Load R into r1:r0
+  poly1305_limbs(R, r0, r1, r1, true);
```

Comment:
What's the intention here when you pass `r1` twice? Just load `R[0]` and `R[2]`. You could use `noreg` to mark an optional operation and check for it in `poly1305_limbs` before loading the corresponding element.

Reply:
ah, I was wondering how to make an 'optional reg' when parameter is not a pointer. `noreg` is exactly what I needed, thanks.